cmake/init-submodule.cmake

```diff
@@ -0,0 +1,40 @@
+# credit: https://cliutils.gitlab.io/modern-cmake/chapters/projects/submodule.html
+
+find_package(Git QUIET)
+
+# NOTE: cannot use PROJECT_SOURCE_DIR here because we need to run this *before*
+#       `project()` is called in order to ensure vcpkg is available.
+
+if(GIT_FOUND AND EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/.git")
+    # Update submodules as needed
+    if(_TILEDB_CMAKE_INIT_GIT_SUBMODULES)
+        message(STATUS "Update submodules")
+        # TODO check pinned version here?
+        execute_process(COMMAND ${GIT_EXECUTABLE} submodule update --init --recursive
+                        WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+                        RESULT_VARIABLE GIT_SUBMOD_RESULT
+                        ECHO_OUTPUT_VARIABLE
+                        COMMAND_ECHO STDOUT)
+        if(NOT GIT_SUBMOD_RESULT EQUAL "0")
+            message(FATAL_ERROR "git submodule update --init --recursive failed with ${GIT_SUBMOD_RESULT}, please checkout submodules")
+        endif()
+    endif()
+endif()
+
+if(NOT EXISTS "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/LICENSE.txt")
+    message(FATAL_ERROR "The submodules were not downloaded! _TILEDB_CMAKE_INIT_GIT_SUBMODULES"
+                        " was turned off or failed. Please update submodules and try again.")
+endif()
+
+#############
+# vcpkg setup
+
+if (TILEDB_GCS)
+    # TODO our_vcpkg_enable
+    list(APPEND VCPKG_MANIFEST_FEATURES "gcs")
+
+    # This must be set *before* `project()`
+    # TODO make sure this works with user-specified toolchain
+    set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake"
+      CACHE STRING "Vcpkg toolchain file")
+endif()
```

Comment:
Would this work instead?

```suggestion
# This must be set *before* `project()`
# TODO make sure this works with user-specified toolchain
set(CMAKE_TOOLCHAIN_FILE "${CMAKE_CURRENT_SOURCE_DIR}/external/vcpkg/scripts/buildsystems/vcpkg.cmake"
    CACHE STRING "Vcpkg toolchain file")

macro(our_vcpkg_enable_if tiledb_feature vcpkg_feature)
    if(${tiledb_feature})
        list(APPEND VCPKG_MANIFEST_FEATURES ${vcpkg_feature})
    endif()
endmacro()

our_vcpkg_enable_if(TILEDB_GCS "gcs")
```

Reply:
I like `our_vcpkg_enable_if` idea. I don't really want to require vcpkg unconditionally just yet.

Reply:
We can use this route and pop in a guard at the top of the file similar to what I've done [here](https://github.com/TileDB-Inc/TileDB/blob/pd/experiment/vcpkg-ports/cmake/Options/TileDBToolchain.cmake#L5-L13). Also note that we should add the check for the super build so we only run vcpkg once.

We should also use that macro. I sketched out support for a few different features and I even noticed it was getting a bit wordy. The macro will help clear that up quite nicely.